File: c++/52_N-Queens_II.cpp
#include <iostream>
#include <vector>
#include <map>
#include <string>
#include <algorithm>

using namespace std;

class Solution {
public:
    int totalNQueens(int n) {
        vector<int> cols;
        int allSol = 0;
        dfs(0, n, cols, allSol);
        return allSol;
    }

    void dfs(int irow, int n, vector<int> cols, int &allSol) {
        if (irow == n) {
            allSol++;
            return;
        }
        for (int icol = 0; icol < n; icol++) {
            if (isValid(irow, icol, cols)) {
                cols.push_back(icol);
                dfs(irow + 1, n, cols, allSol);
                cols.pop_back();
            }
        }

    }
    bool isValid(int irow, int icol, vector<int> cols) {
        for (int row = 0; row < cols.size(); row++) {
            if (icol == cols[row] or abs(irow - row) == abs(icol - cols[row]))
                return false;
        }
        return true;
    }
};



int main() {
	Solution s;
}

File: python/52_N-Queens_II.py
#!/usr/bin/python
import unittest



class Solution(object):
    def totalNQueens(self, n):
        def isValid(irow, icol, cols):
            for row, col in enumerate(cols):
                if icol == col or abs(irow - row) == abs(icol - col):
                    return False
            return True
        def dfs(irow, cols):
            if irow == n:
                allSol[0] += 1
                return
            for icol in range(n):
                if isValid(irow, icol, cols):
                    dfs(irow + 1, cols + [icol])

        allSol = [0]
        dfs(0, [])
        return allSol[0]




def main():
    s = Solution()
    print s.totalNQueens(4)


if __name__ == "__main__":
    main()
